aurora/core.py

```diff
@@ -198,19 +239,197 @@ def setup_grids(self):
             # use rho_vol = rvol/rvol_lcfs
             self.rhop_grid = self.rvol_grid / self.rvol_lcfs
 
-        # ----------------
+        # possibly adapt the namelist 'timing' for ELMs, before creating the time grid, if asked to do so
+        if self.namelist['ELM_model']['ELM_flag'] and self.namelist['ELM_model']['adapt_time_grid']:
+            
+            # updated version of the namelist 'timing
+            self.namelist['timing'] = grids_utils.ELM_time_grid(self.namelist['timing'], self.namelist['ELM_model'])
+
         # define time grid ('timing' must be in namelist)
         self.time_grid, self.save_time = grids_utils.create_time_grid(
             timing=self.namelist["timing"], plot=False
         )
         self.time_out = self.time_grid[self.save_time]
+        self.time_plot = self.time_out + self.namelist["timing"]["time_start_plot"]
 
         # create array of 0's of length equal to self.time_grid, with 1's where sawteeth must be triggered
         self.saw_on = np.zeros_like(self.time_grid)
         input_saw_times = self.namelist["saw_model"]["times"]
         self.saw_times = np.array(input_saw_times)[input_saw_times < self.time_grid[-1]]
         if self.namelist["saw_model"]["saw_flag"] and len(self.saw_times) > 0:
             self.saw_on[self.time_grid.searchsorted(self.saw_times)] = 1
+            
+        # calculate core plasma volume (i.e. volume of region until rvol = rvol_lcfs) and total plasma volume
+        ones = np.ones((len(self.time_out),len(self.rvol_grid)))
+        self.core_vol = grids_utils.vol_int(ones[:,:],self.rvol_grid,self.pro_grid,self.Raxis_cm,rvol_max=self.rvol_lcfs)[0]
+        self.plasma_vol = grids_utils.vol_int(ones[:,:],self.rvol_grid,self.pro_grid,self.Raxis_cm,rvol_max=None)[0]
+                 
+        
+    def setup_PWI_model(self):
+        """Method to set up Aurora inputs related to the advanced plasma-wall interaction model.
+        """
+        
+        # effective wall surfaces, accounting for roughness
+        self.surf_mainwall_eff = self.surf_mainwall*self.mainwall_roughness #cm^2
+        self.surf_divwall_eff = self.surf_divwall*self.divwall_roughness #cm^2
+
+        if self.advanced_PWI_flag: # advanced PWI model used: calculate parameters for fortran routine
+        
+            # Data for advanced PWI model for now available only for He as impurity and W as wall material
+           
+            if self.imp != 'He':
+                raise ValueError(f"Advanced PWI model not available for impurity {self.imp}!")
+            
+            # keys for fortran routine
+            if self.main_wall_material == 'W':
```

Comment:
@antonellozito is working on AUG, so has so far included TRIM data only for W, but it would make a lot of sense to include datasets for C too. @antonellozito would you be able to handle that?

Reply:
@odstrcilt Well, it works for tungsten (and, I add, only for He as impurity). Datasets for other combination wall material/impurity should (and can) be created, which shouldn't be a problem. Next days I will talk with our colleague from the material group who created the dataset for He-W, and I will ask him to pass me the TRIM input files he used, so I will be able to personally add, in the future, datasets for more wall materials and more impurities.